core/blockchain.go

```diff
@@ -531,7 +526,7 @@ func (bc *BlockChain) HasBlockAndState(hash common.Hash) bool {
 		return false
 	}
 	// Ensure the associated state is also present
-	_, err := state.New(block.Root(), bc.chainDb)
+	_, err := state.New(block.Root(), bc.stateCache)
```

Comment:
The `state.New` does a lot of allocation and things (https://github.com/ethereum/go-ethereum/blob/master/core/state/statedb.go#L89) . Seems a bit wasteful to instantiate one and just throw it away. Couldn't we just verify the state without actually allocating the entire instance in this case?